From 09e13163407151c280be4a5d51dd3b57aca05089 Mon Sep 17 00:00:00 2001
Message-Id: <09e13163407151c280be4a5d51dd3b57aca05089.1425610844.git.chang-joon.lee@intel.com>
In-Reply-To: <6fa2ab23adf6d7de781e906d04a0224410262d9b.1425610844.git.chang-joon.lee@intel.com>
References: <6fa2ab23adf6d7de781e906d04a0224410262d9b.1425610844.git.chang-joon.lee@intel.com>
From: Durgadoss R <durgadoss.r@intel.com>
Date: Thu, 5 Feb 2015 18:20:34 +0530
Subject: [PATCH 03/27] MUST_REBASE [VPG]: drivers/video/adf: Add pipe offsets
 for M2/N2

This patch adds:
* Required pipe offsets for m2_n2 registers programming
* A method to program m2_n2
* Macro that defines the switch to lower refresh rate in eDP.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5760
Change-Id: I550134adbdd2a2f395b3e98300905cf78b849b12
Signed-off-by: Durgadoss R <durgadoss.r@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_pipe.c        |   13 +++++++++++++
 .../video/adf/intel/include/core/vlv/vlv_dc_regs.h |    1 +
 .../video/adf/intel/include/core/vlv/vlv_pipe.h    |    8 ++++++++
 3 files changed, 22 insertions(+)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_pipe.c b/drivers/video/adf/intel/core/vlv/vlv_pipe.c
index cc1d11f..884cfd9 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_pipe.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_pipe.c
@@ -27,6 +27,14 @@ void vlv_pipe_program_m_n(struct vlv_pipe *pipe, struct intel_link_m_n *m_n)
 	REG_WRITE(pipe->linkn1_offset, m_n->link_n);
 }
 
+void vlv_pipe_program_m2_n2(struct vlv_pipe *pipe, struct intel_link_m_n *m_n)
+{
+	REG_WRITE(pipe->datam2_offset, TU_SIZE(m_n->tu) | m_n->gmch_m);
+	REG_WRITE(pipe->datan2_offset, m_n->gmch_n);
+	REG_WRITE(pipe->linkm2_offset, m_n->link_m);
+	REG_WRITE(pipe->linkn2_offset, m_n->link_n);
+}
+
 bool vlv_pipe_vblank_on(struct vlv_pipe *pipe)
 {
 	u32 val = REG_READ(pipe->status_offset);
@@ -410,6 +418,11 @@ bool vlv_pipe_init(struct vlv_pipe *pipe, enum pipe pipeid)
 	pipe->linkm1_offset = PIPE_LINK_M1(pipeid);
 	pipe->linkn1_offset = PIPE_LINK_N1(pipeid);
 
+	pipe->datam2_offset = PIPE_DATA_M2(pipeid);
+	pipe->datan2_offset = PIPE_DATA_N2(pipeid);
+	pipe->linkm2_offset = PIPE_LINK_M2(pipeid);
+	pipe->linkn2_offset = PIPE_LINK_N2(pipeid);
+
 	pipe->psr_ctrl_offset = VLV_EDP_PSR_CONTROL(pipeid);
 	pipe->psr_sts_offset = VLV_EDP_PSR_STATUS(pipeid);
 	pipe->psr_crc1_offset = VLV_EDP_PSR_CRC1(pipeid);
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
index 53d1e9b..ecd253d 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
@@ -3574,6 +3574,7 @@ extern int chv_cursor_offsets[];
 #define   PIPECONF_DITHER_TYPE_ST1 (1<<2)
 #define   PIPECONF_DITHER_TYPE_ST2 (2<<2)
 #define   PIPECONF_DITHER_TYPE_TEMP (3<<2)
+#define   PIPECONF_EDP_RR_SWITCH_VLV (1 << 14)
 #define _PIPEASTAT		(VLV_DISPLAY_BASE + 0x70024)
 #define   PIPE_FIFO_UNDERRUN_STATUS		(1UL<<31)
 #define   SPRITE1_FLIPDONE_INT_EN_VLV		(1UL<<30)
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_pipe.h b/drivers/video/adf/intel/include/core/vlv/vlv_pipe.h
index 8df11a0..9cb640b 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_pipe.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_pipe.h
@@ -36,6 +36,12 @@ struct vlv_pipe {
 	u32 linkm1_offset;
 	u32 linkn1_offset;
 
+	/* m2_n2 values for eDP DRRS */
+	u32 datam2_offset;
+	u32 datan2_offset;
+	u32 linkm2_offset;
+	u32 linkn2_offset;
+
 	u32 src_size_offset;
 
 	/* eDP PSR related registers */
@@ -49,6 +55,8 @@ struct vlv_pipe {
 
 void vlv_pipe_program_m_n(struct vlv_pipe *pipe,
 		struct intel_link_m_n *m_n);
+void vlv_pipe_program_m2_n2(struct vlv_pipe *pipe,
+		struct intel_link_m_n *m_n);
 bool vlv_pipe_vblank_on(struct vlv_pipe *pipe);
 bool vlv_pipe_vblank_off(struct vlv_pipe *pipe);
 bool vlv_pipe_wait_for_vblank(struct vlv_pipe *pipe);
-- 
1.7.9.5

